plugins/jenkins/app/models/samson/jenkins.rb

```diff
@@ -30,17 +52,142 @@ def initialize(job_name, deploy)
+      present_build_params = build_params(conf).map(&:to_sym)
+      missing_build_params = expected_build_parameters.to_set - present_build_params.to_set
+      if missing_build_params.any?
+        add_build_parameters(conf, missing_build_params)
+        post_config = true
+      end
+      unless description_exists?(conf)
+        add_job_description(conf)
+        post_config = true
+      end
+      if post_config
+        post_job_config(conf)
+      end
+    end
+
+    def build_params(conf)
+      params = conf.xpath("//hudson.model.StringParameterDefinition")
+      params_array = []
+      params.each do |param|
+        param.children.each do |value|
+          if value.name == "name"
+            if val = value.content.split(JENKINS_BUILD_PARAMETRS_PREFIX, 2)[1]
+              params_array << val
+            end
+          end
+        end
+      end
+      params_array
+    end
+
+    def description(conf)
+      conf.xpath("//description").first
+    end
+
+    def jenkins_desc_job_name
+      '* ' + deploy.project.name + ' - ' + deploy.stage.name
+    end
+
+    def add_job_description(conf)
+      # get previous content and job-name from desc
+      job_names = []
+      prev_content = []
+      desc = description(conf)
+      content = desc.content.split("\n").map(&:squish)
+      desc_start_index = content.index(JENKINS_DESC_HEADER)
+      desc_end_index = content.index(JENKINS_DESC_FOOTER)
+      if desc_start_index && desc_end_index
+        job_names = content[desc_start_index..desc_end_index].select { |c| c.starts_with?('*') }
+        prev_content = content[0...desc_start_index] + content[(desc_end_index + 1)..content.size]
+      else
+        prev_content = content
+      end
+
+      # add present job name if not included
+      unless job_names.include?(jenkins_desc_job_name)
+        job_names.append(jenkins_desc_job_name)
+      end
+
+      # update desc
+      desc.content = [
+        prev_content.flatten,
+        JENKINS_DESC_HEADER,
+        JENKINS_JOB_DESC % {job_names: job_names.join("\n")},
+        JENKINS_DESC_FOOTER
+      ].join("\n")
+    end
+
+    def description_exists?(conf)
+      job_name = jenkins_desc_job_name + "\n"
+      desc = description(conf).content
+      desc.include?(JENKINS_DESC_HEADER) &&
+      desc.include?(JENKINS_DESC_FOOTER) &&
+      desc.include?(job_name)
+    end
+
+    def find_or_add_parameter_definition(conf)
+      properties_element = conf.xpath("//parameterDefinitions").first
+      unless properties_element
+        conf.xpath("//properties").first.add_child(
+          '<hudson.model.ParametersDefinitionProperty>'\
+          '<parameterDefinitions />'\
+          '</hudson.model.ParametersDefinitionProperty>'
+        )
+      end
+    end
+
+    def add_build_parameters(conf, missing_parameters)
+      find_or_add_parameter_definition(conf)
+      properties_element = conf.xpath("//parameterDefinitions").first
```

Comment:
parameterDefinitions is duplicated now, find_or_add_parameter_definition can own the whole concern -> 
```
properties_element = find_or_add_parameter_definition(conf)
```